### ISSUE TITLE
[ADX趋势强度策略] | ADX Trend Strength Strategy

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4476]
**标签**: quantitative-strategy
### 摘要
利用平均方向指数（ADX）量化趋势强度，通过DI+/DI-交叉信号识别趋势方向，构建多时间框架的趋势跟踪策略。

### 算法描述

**核心原理：**

ADX（Average Directional Index）是J. Welles Wilder Jr.发明的趋势强度指标，用于衡量趋势的强弱程度而非方向。ADX读数越高表示趋势越强，低于20表示市场处于区间震荡。

**数学模型：**

1. **方向移动（Directional Movement）**：
$$+DM = \begin{cases} High_t - High_{t-1} & \text{if } High_t > High_{t-1} \text{ and } High_t - High_{t-1} > Low_{t-1} - Low_t \\ 0 & \text{otherwise} \end{cases}$$

$$-DM = \begin{cases} Low_{t-1} - Low_t & \text{if } Low_{t-1} > Low_t \text{ and } Low_{t-1} - Low_t > High_t - High_{t-1} \\ 0 & \text{otherwise} \end{cases}$$

2. **真实波幅（True Range）**：
$$TR = \max(H_t - L_t, |H_t - C_{t-1}|, |L_t - C_{t-1}|)$$

3. **方向指标（Directional Indicators）**：
$$+DI = \frac{EMA(+DM, period)}{ATR} \times 100$$
$$-DI = \frac{EMA(-DM, period)}{ATR} \times 100$$

4. **平均方向指数（ADX）**：
$$ADX = \frac{EMA(|+DI - (-DI)|}{+DI + (-DI)}, period) \times 100$$

5. **趋势强度分级**：
- $ADX < 20$：无趋势（区间震荡）
- $20 \leq ADX < 40$：弱趋势
- $40 \leq ADX < 60$：中等趋势
- $ADX \geq 60$：强趋势

6. **DI交叉信号**：
$$Signal = \begin{cases} \text{BUY} & \text{if } +DI > -DI \text{ and } ADX > 20 \\ \text{SELL} & \text{if } -DI > +DI \text{ and } ADX > 20 \\ \text{HOLD} & \text{if } ADX \leq 20 \end{cases}$$

**交易策略：**

1. **趋势跟踪**：ADX > 25 且 +DI 上穿 -DI → 买入；ADX > 25 且 -DI 上穿 +DI → 卖出
2. **趋势强度过滤**：ADX < 20 时不交易，等待趋势确立
3. **止损优化**：使用 ATR 倍数作为动态止损
4. **多时间框架确认**：日线 ADX > 30 且周线 ADX 上升 → 确认趋势

### Python代码示例

```python
"""
ADX趋势强度策略
ADX Trend Strength Strategy

核心逻辑：
1. 计算ADX及DI+/DI-指标
2. 趋势强度分级与信号生成
3. 多时间框架确认
4. ATR动态止损

作者：Ginkgo量化研究组
版本：v1.0
"""

import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from enum import Enum


class TrendStrength(Enum):
    """趋势强度等级"""
    NO_TREND = "no_trend"           # ADX < 20
    WEAK = "weak"                  # 20 <= ADX < 40
    MODERATE = "moderate"          # 40 <= ADX < 60
    STRONG = "strong"              # ADX >= 60


class ADXCalculator:
    """ADX指标计算器"""
    
    def __init__(self, period: int = 14):
        """
        初始化计算器
        
        Args:
            period: 计算周期（通常14）
        """
        self.period = period
    
    def calculate_true_range(
        self,
        high: pd.Series,
        low: pd.Series,
        close: pd.Series
    ) -> pd.Series:
        """
        计算真实波幅
        
        Args:
            high: 最高价
            low: 最低价
            close: 收盘价
        
        Returns:
            TR序列
        """
        tr1 = high - low
        tr2 = abs(high - close.shift(1))
        tr3 = abs(low - close.shift(1))
        
        tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)
        return tr
    
    def calculate_directional_movement(
        self,
        high: pd.Series,
        low: pd.Series
    ) -> Tuple[pd.Series, pd.Series]:
        """
        计算方向移动
        
        Args:
            high: 最高价
            low: 最低价
        
        Returns:
            (+DM, -DM)
        """
        high_diff = high.diff()
        low_diff = -low.diff()
        
        plus_dm = pd.Series(0.0, index=high.index)
        minus_dm = pd.Series(0.0, index=high.index)
        
        # +DM
        up_move = high_diff.where(high_diff > low_diff, 0)
        down_move = low_diff.where(low_diff > high_diff, 0)
        
        plus_dm = up_move.where(up_move > down_move, 0)
        minus_dm = down_move.where(down_move > up_move, 0)
        
        return plus_dm, minus_dm
    
    def calculate_adx(
        self,
        high: pd.Series,
        low: pd.Series,
        close: pd.Series
    ) -> pd.DataFrame:
        """
        计算ADX系列指标
        
        Args:
            high: 最高价
            low: 最低价
            close: 收盘价
        
        Returns:
            包含ADX、DI+、DI-的DataFrame
        """
        # 真实波幅
        tr = self.calculate_true_range(high, low, close)
        atr = tr.ewm(span=self.period, adjust=False).mean()
        
        # 方向移动
        plus_dm, minus_dm = self.calculate_directional_movement(high, low)
        
        # 平滑DM
        plus_dm_smooth = plus_dm.ewm(span=self.period, adjust=False).mean()
        minus_dm_smooth = minus_dm.ewm(span=self.period, adjust=False).mean()
        
        # DI+和DI-
        di_plus = 100 * plus_dm_smooth / atr
        di_minus = 100 * minus_dm_smooth / atr
        
        # DX
        dx = 100 * abs(di_plus - di_minus) / (di_plus + di_minus)
        
        # ADX
        adx = dx.ewm(span=self.period, adjust=False).mean()
        
        return pd.DataFrame({
            'ADX': adx,
            'DI+': di_plus,
            'DI-': di_minus
        })
    
    def detect_trend_strength(self, adx: float) -> TrendStrength:
        """
        检测趋势强度等级
        
        Args:
            adx: ADX值
        
        Returns:
            趋势强度等级
        """
        if adx < 20:
            return TrendStrength.NO_TREND
        elif adx < 40:
            return TrendStrength.WEAK
        elif adx < 60:
            return TrendStrength.MODERATE
        else:
            return TrendStrength.STRONG


class ADXCrossSignal:
    """DI交叉信号检测"""
    
    def __init__(self, adx_threshold: float = 20.0):
        """
        初始化
        
        Args:
            adx_threshold: ADX阈值（低于此值认为无趋势）
        """
        self.adx_threshold = adx_threshold
    
    def detect_cross(
        self,
        di_plus: float,
        di_minus: float,
        prev_di_plus: float,
        prev_di_minus: float
    ) -> Tuple[str, str]:
        """
        检测DI交叉
        
        Args:
            di_plus: 当前DI+
            di_minus: 当前DI-
            prev_di_plus: 前一期DI+
            prev_di_minus: 前一期DI-
        
        Returns:
            (交叉类型, 交叉强度)
        """
        # 金叉：DI+从下穿越DI-
        if prev_di_plus <= prev_di_minus and di_plus > di_minus:
            return "BUY", min(1.0, (di_plus - di_minus) / 50)
        
        # 死叉：DI-从下穿越DI+
        if prev_di_minus <= prev_di_plus and di_minus > di_plus:
            return "SELL", min(1.0, (di_minus - di_plus) / 50)
        
        return "HOLD", "0"
    
    def filter_by_adx(
        self,
        signal: str,
        adx: float,
        min_adx: float = 25.0
    ) -> Tuple[str, float]:
        """
        根据ADX过滤信号
        
        Args:
            signal: 原始信号
            adx: 当前ADX值
            min_adx: 最小ADX要求
        
        Returns:
            (过滤后信号, ADX强度因子)
        """
        adx_factor = min(adx / 100, 1.0)
        
        if adx < min_adx:
            return "HOLD", adx_factor
        
        return signal, adx_factor


class ATRStopLoss:
    """ATR动态止损"""
    
    def __init__(self, atr_multiplier: float = 2.0):
        """
        初始化
        
        Args:
            atr_multiplier: ATR倍数
        """
        self.atr_multiplier = atr_multiplier
    
    def calculate_stop_loss(
        self,
        entry_price: float,
        atr: float,
        direction: str,
        highest_since_entry: float = None,
        lowest_since_entry: float = None
    ) -> Tuple[float, float]:
        """
        计算止损位
        
        Args:
            entry_price: 入场价格
            atr: ATR值
            direction: 方向（'LONG'或'SHORT'）
            highest_since_entry: 入场后最高价
            lowest_since_entry: 入场后最低价
        
        Returns:
            (止损价, 追踪止损价)
        """
        atr_value = atr * self.atr_multiplier
        
        if direction == "LONG":
            stop_loss = entry_price - atr_value
            
            # 追踪止损（基于最高点）
            if highest_since_entry is not None:
                trailing_stop = highest_since_entry - atr_value
                stop_loss = max(stop_loss, trailing_stop)
        else:
            stop_loss = entry_price + atr_value
            
            if lowest_since_entry is not None:
                trailing_stop = lowest_since_entry + atr_value
                stop_loss = min(stop_loss, trailing_stop)
        
        return stop_loss, stop_loss


class ADXTrendStrengthStrategy:
    """ADX趋势强度策略"""
    
    def __init__(
        self,
        adx_period: int = 14,
        adx_threshold: float = 20.0,
        min_adx_for_entry: float = 25.0,
        atr_multiplier: float = 2.0,
        use_multi_timeframe: bool = True
    ):
        """
        初始化策略
        
        Args:
            adx_period: ADX计算周期
            adx_threshold: ADX阈值
            min_adx_for_entry: 入场最小ADX要求
            atr_multiplier: ATR止损倍数
            use_multi_timeframe: 是否使用多时间框架
        """
        self.adx_period = adx_period
        self.adx_threshold = adx_threshold
        self.min_adx_for_entry = min_adx_for_entry
        self.atr_multiplier = atr_multiplier
        self.use_multi_timeframe = use_multi_timeframe
        
        self.adx_calculator = ADXCalculator(adx_period)
        self.cross_signal = ADXCrossSignal(adx_threshold)
        self.atr_stop = ATRStopLoss(atr_multiplier)
    
    def calculate_indicators(self, data: pd.DataFrame) -> pd.DataFrame:
        """
        计算所有指标
        
        Args:
            data: 价格数据
        
        Returns:
            指标数据
        """
        adx_df = self.adx_calculator.calculate_adx(
            data['high'],
            data['low'],
            data['close']
        )
        
        # ATR用于止损
        tr = self.adx_calculator.calculate_true_range(
            data['high'], data['low'], data['close']
        )
        adx_df['ATR'] = tr.ewm(span=self.adx_period, adjust=False).mean()
        
        return adx_df
    
    def get_daily_and_weekly_adx(
        self,
        daily_data: pd.DataFrame
    ) -> Tuple[float, float]:
        """
        获取日线和周线ADX
        
        Args:
            daily_data: 日线数据
        
        Returns:
            (日线ADX, 周线ADX)
        """
        daily_adx = self.calculate_indicators(daily_data)['ADX'].iloc[-1]
        
        # 模拟周线数据（每周5个交易日）
        weekly_data = daily_data.iloc[::5].copy()
        weekly_adx = self.adx_calculator.calculate_adx(
            weekly_data['high'],
            weekly_data['low'],
            weekly_data['close']
        )['ADX'].iloc[-1]
        
        return daily_adx, weekly_adx
    
    def run_backtest(
        self,
        daily_data: pd.DataFrame,
        weekly_data: Optional[pd.DataFrame] = None
    ) -> pd.DataFrame:
        """
        回测策略
        
        Args:
            daily_data: 日线数据
            weekly_data: 周线数据（可选）
        
        Returns:
            回测结果
        """
        results = []
        position = None
        entry_date = None
        entry_price = 0
        highest_since_entry = 0
        lowest_since_entry = float('inf')
        
        # 计算指标
        indicators = self.calculate_indicators(daily_data)
        adx_values = indicators['ADX']
        di_plus = indicators['DI+']
        di_minus = indicators['DI-']
        atr_values = indicators['ATR']
        
        for i in range(self.adx_period, len(daily_data)):
            date = daily_data['date'].iloc[i]
            close = daily_data['close'].iloc[i]
            high = daily_data['high'].iloc[i]
            low = daily_data['low'].iloc[i]
            
            current_adx = adx_values.iloc[i]
            current_di_plus = di_plus.iloc[i]
            current_di_minus = di_minus.iloc[i]
            current_atr = atr_values.iloc[i]
            
            prev_di_plus = di_plus.iloc[i-1]
            prev_di_minus = di_minus.iloc[i-1]
            
            if position is None:
                # 检测DI交叉
                cross_type, cross_strength = self.cross_signal.detect_cross(
                    current_di_plus, current_di_minus,
                    prev_di_plus, prev_di_minus
                )
                
                # 过滤信号
                signal, adx_factor = self.cross_signal.filter_by_adx(
                    cross_type, current_adx, self.min_adx_for_entry
                )
                
                # 多时间框架确认
                if self.use_multi_timeframe and weekly_data is not None:
                    daily_adx, weekly_adx = self.get_daily_and_weekly_adx(
                        daily_data.iloc[:i+1]
                    )
                    if signal == "BUY" and weekly_adx < daily_adx:
                        signal = "HOLD"
                    elif signal == "SELL" and weekly_adx < daily_adx:
                        signal = "HOLD"
                
                confidence = cross_strength * adx_factor
                
                if signal in ["BUY", "SELL"]:
                    position = {
                        "type": "LONG" if signal == "BUY" else "SHORT",
                        "entry_date": date,
                        "entry_price": close,
                        "entry_adx": current_adx,
                        "entry_di_plus": current_di_plus,
                        "entry_di_minus": current_di_minus
                    }
                    highest_since_entry = high
                    lowest_since_entry = low
                    
                    results.append({
                        "date": date,
                        "action": signal,
                        "price": close,
                        "adx": current_adx,
                        "di_plus": current_di_plus,
                        "di_minus": current_di_minus,
                        "atr": current_atr,
                        "confidence": confidence,
                        "trend_strength": self.adx_calculator.detect_trend_strength(current_adx).value,
                        "pnl": 0
                    })
            else:
                # 更新高低价
                highest_since_entry = max(highest_since_entry, high)
                lowest_since_entry = min(lowest_since_entry, low)
                
                # 计算盈亏
                if position['type'] == "LONG":
                    pnl = (close - entry_price) / entry_price
                    stop_loss, _ = self.atr_stop.calculate_stop_loss(
                        entry_price, current_atr, "LONG",
                        highest_since_entry, lowest_since_entry
                    )
                else:
                    pnl = (entry_price - close) / entry_price
                    stop_loss, _ = self.atr_stop.calculate_stop_loss(
                        entry_price, current_atr, "SHORT",
                        highest_since_entry, lowest_since_entry
                    )
                
                should_close = False
                close_reason = ""
                
                # 止损检查
                if position['type'] == "LONG" and close <= stop_loss:
                    should_close = True
                    close_reason = "stop_loss"
                elif position['type'] == "SHORT" and close >= stop_loss:
                    should_close = True
                    close_reason = "stop_loss"
                
                # 趋势反转
                prev_adx = adx_values.iloc[i-1]
                if current_adx > 20 and prev_adx <= prev_di_minus and current_di_minus > current_di_plus:
                    should_close = True
                    close_reason = "trend_reversal"
                
                # ADX大幅下降
                if current_adx < 15:
                    should_close = True
                    close_reason = "adx_collapse"
                
                if should_close:
                    results.append({
                        "date": date,
                        "action": f"CLOSE({close_reason})",
                        "price": close,
                        "adx": current_adx,
                        "atr": current_atr,
                        "pnl": pnl,
                        "holding_days": (date - entry_date).days
                    })
                    position = None
                    entry_date = None
        
        return pd.DataFrame(results)


def generate_synthetic_trend_data(days: int = 252) -> pd.DataFrame:
    """生成合成趋势数据"""
    np.random.seed(42)
    
    dates = pd.date_range(start='2024-01-01', periods=days, freq='D')
    
    # 模拟趋势市场
    prices = [100]
    for i in range(days):
        if i < 50:
            ret = np.random.normal(0.003, 0.015)  # 上涨趋势
        elif i < 100:
            ret = np.random.normal(-0.001, 0.02)  # 震荡
        elif i < 150:
            ret = np.random.normal(0.004, 0.02)  # 上涨
        elif i < 200:
            ret = np.random.normal(-0.003, 0.025)  # 下跌
        else:
            ret = np.random.normal(0.001, 0.015)  # 震荡
        
        prices.append(prices[-1] * (1 + ret))
    
    prices = prices[1:]
    
    data = pd.DataFrame({
        'date': dates,
        'close': prices,
        'open': [p * (1 + np.random.uniform(-0.005, 0.005)) for p in prices],
        'high': [p * (1 + np.random.uniform(0, 0.015)) for p in prices],
        'low': [p * (1 - np.random.uniform(0, 0.015)) for p in prices],
        'volume': np.random.randint(1000000, 10000000, days)
    })
    
    return data


if __name__ == "__main__":
    print("=" * 60)
    print("ADX趋势强度策略回测")
    print("=" * 60)
    
    # 生成数据
    daily_data = generate_synthetic_trend_data(days=252)
    
    # 创建策略
    strategy = ADXTrendStrengthStrategy(
        adx_period=14,
        adx_threshold=20.0,
        min_adx_for_entry=25.0,
        atr_multiplier=2.0
    )
    
    # 回测
    results = strategy.run_backtest(daily_data)
    
    if len(results) > 0:
        print("\n【回测结果】")
        print(results)
        
        close_results = results[results['action'].str.startswith('CLOSE')]
        if len(close_results) > 0:
            print(f"\n总交易次数: {len(results)}")
            print(f"买入次数: {len(results[results['action'] == 'BUY'])}")
            print(f"卖出次数: {len(results[results['action'] == 'SELL'])}")
            print(f"平均持仓天数: {close_results['holding_days'].mean():.1f}")
            print(f"胜率: {(close_results['pnl'] > 0).mean()*100:.1f}%")
    else:
        print("\n未产生交易信号")
